DeepSpeech.py

```diff
@@ -1445,82 +1453,95 @@ def end(self, session):
-                        collect_results(report_results, batch_report[0])
-                        # Add batch to total_mean_edit_distance
-                        total_mean_edit_distance += batch_report[1]
-
-                # Gathering job results
-                job.loss = total_loss / job.steps
-                if job.report:
-                    job.mean_edit_distance = total_mean_edit_distance / job.steps
-                    job.wer, job.samples = calculate_report(report_results)
-
-                # Send the current job to coordinator and receive the next one
-                log_debug('Sending %s...' % str(job))
-                job = COORD.next_job(job)
+                        # Reset mean edit distance
+                        total_mean_edit_distance = 0.0
+                        # Create report results tuple
+                        report_results = ([],[],[],[])
+                        # Extend the session.run parameters
+                        report_params = [results_tuple, mean_edit_distance]
+                    else:
+                        report_params = []
+
+                    # So far the only extra parameter is the feed_dict
+                    extra_params = { 'feed_dict': feed_dict }
+
+                    # Loop over the batches
+                    for job_step in range(job.steps):
+                        if session.should_stop():
+                            break
+
+                        log_debug('Starting batch...')
+                        # Compute the batch
+                        _, current_step, batch_loss, batch_report = session.run([train_op, global_step, loss, report_params], **extra_params)
+
+                        # Uncomment the next line for debugging race conditions / distributed TF
+                        log_debug('Finished batch step %d.' % current_step)
+
+                        # Add batch to loss
+                        total_loss += batch_loss
+
+                        if job.report:
+                            # Collect individual sample results
+                            collect_results(report_results, batch_report[0])
+                            # Add batch to total_mean_edit_distance
+                            total_mean_edit_distance += batch_report[1]
+
+                    # Gathering job results
+                    job.loss = total_loss / job.steps
+                    if job.report:
+                        job.mean_edit_distance = total_mean_edit_distance / job.steps
+                        job.wer, job.samples = calculate_report(report_results)
+
+                    # Send the current job to coordinator and receive the next one
+                    log_debug('Sending %s...' % str(job))
+                    job = COORD.next_job(job)
+            except Exception as e:
+                log_error(e)
+                # Calling all hook's end() methods to end blocking calls
+                for hook in hooks:
+                    hook.end(session)
```

Comment:
@kdavis-mozilla Against my guesses this seems to work and therefore I prefer to call every hook's `end()` - like in your first suggestion.